src/main/portfolio_driver.cpp

```diff
@@ -0,0 +1,531 @@
+      // 0 = solved, 1 = not solved
+      SolveStatus rc = SolveStatus::STATUS_UNSOLVED;
+      if (d_ctx.solveCommands(d_commands))
+      {
+        Result res = d_ctx.d_executor->getResult();
+        if (res.isSat() || res.isUnsat())
+        {
+          rc = SolveStatus::STATUS_SOLVED;
+        }
+      }
+      _exit(rc);
+    }
+    job.d_errPipe.closeIn();
+    job.d_outPipe.closeIn();
+
+    // Start the timeout process
+    if (d_timeout > 0 && job.d_config.d_timeout > 0)
+    {
+      job.d_timeout = fork();
+      if (job.d_timeout == 0)
+      {
+        auto duration = std::chrono::duration<double, std::milli>(
+            job.d_config.d_timeout * d_timeout);
+        std::this_thread::sleep_for(duration);
+        kill(job.d_worker, SIGKILL);
+        _exit(0);
+      }
+    }
+
+    ++d_nextJob;
+    ++d_running;
+    job.d_state = JobState::RUNNING;
+  }
+
+  /**
+   * Check whether some process terminated and solved the input. If so,
+   * forward the child process output to the main out and return true.
+   * Otherwise return false.
+   * If child and status are given, only the job with this particular worker is
+   * considered and status is assumed to be the wstatus reported by waitpid.
+   */
+  bool checkResults(pid_t child = -1, int status = 0)
+  {
+    // check d_jobs for items where worker has terminated and timeout != -1
+    for (auto& job : d_jobs)
+    {
+      // has not been started yet
+      if (job.d_state == JobState::PENDING) continue;
+      // has already been analyzed
+      if (job.d_state == JobState::DONE) continue;
+      // was given an explicit child, but this is not it.
+      if (child != -1 && job.d_worker != child) continue;
+
+      int wstatus = 0;
+      pid_t res = 0;
+      if (child == -1)
+      {
+        res = waitpid(job.d_worker, &wstatus, WNOHANG);
+        // has not terminated yet
+        if (res == 0) continue;
+        if (res == -1) continue;
+      }
+      else
+      {
+        res = child;
+        wstatus = status;
+      }
+      // mark as analyzed
+      Trace("portfolio") << "Finished " << job.d_config << std::endl;
+      job.d_state = JobState::DONE;
+      --d_running;
+      // check if exited normally
+      if (WIFSIGNALED(wstatus))
+      {
+        continue;
+      }
+      if (WIFEXITED(wstatus))
+      {
+        if (WEXITSTATUS(wstatus) == SolveStatus::STATUS_SOLVED)
+        {
+          Trace("portfolio") << "Successful!" << std::endl;
+          job.d_errPipe.flushTo(std::cerr);
+          job.d_outPipe.flushTo(std::cout);
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  ExecutionContext& d_ctx;
+  std::vector<std::unique_ptr<cvc5::Command>> d_commands;
+  /**
+   * All jobs.
+   */
+  std::vector<Job> d_jobs;
+  /** The id of the next job to be started within d_jobs */
+  size_t d_nextJob = 0;
+  /** The number of currently running jobs */
+  size_t d_running = 0;
+  const uint64_t d_numJobs;
```

Comment:
This is the maximum number of parallel jobs and not `d_jobs.size()`, right? Maybe rename to `d_maxJobs`?

Reply:
done